### PR TITLE
Extract critical css to head for amp package

### LINKED ISSUE #559: Move all the CSS to the <head>. [5pt]
This is part of the [AMP feature](https://community.frontity.org/t/amp-package/388/). For full context please check out the [final Implementation proposal](https://community.frontity.org/t/amp-package/388/10?u=santosguillamot).

---

The Google AMP framework requires that all the CSS is placed in a `<style>` tag of the head.

For that reason, the `@frontity/amp` package needs to use Emotion's `extractCritical` API and move the extracted CSS to the `<head>`.

This can be done with a `beforeSSR` function. It'll be necessary to create a separate `server.js` file because we don't want to add `create-emotion-server` and `react-dom/server` to the client bundle.

```js
// packages/amp/src/server.js
import { CacheProvider } from "@emotion/core";
import createEmotionServer from "create-emotion-server";
import createCache from "@emotion/cache";
import { renderToStaticMarkup } from "react-dom/server";
import ampTemplate from "./templates/amp";

export default {
  actions: {
    amp: {
      beforeSSR: ({ libraries }) => {
        // Create an Emotion instance for this SSR.
        const cache = createCache();
        const { extractCritical } = createEmotionServer(cache);

        libraries.frontity.render = (App) => {
          // Add the Provider to the App.
          App = (
            <CacheProvider value={cache}>
              <App />
            </CacheProvider>
          );

          // Run extract critical.
          const { html, css, ids } = extractCritical(
            libraries.frontity.render(App)
          );

          // Replace the template injecting the CSS in the `<head>`.
          libraries.frontity.template = ampTemplate(css, ids);

          return html;
        };
      },
    },
  },
};
```

We also need to be able to add the `CacheProvider` to the client. That can be done with a `beforeCSR` function.

```js
// packages/amp/src/client.js
import createCache from "@emotion/cache";

export default {
  actions: {
    amp: {
      beforeCSR: ({ libraries }) => {
        const cache = createCache();

        libraries.frontity.App = () => (
          <CacheProvider value={cache}>
            <libraries.frontity.App />
          </CacheProvider>
        );
      },
    },
  },
};
```

For context, this is the way Emotion worked in Frontity at first. This is the final commit, including a fix for the `Global` component: https://github.com/frontity/frontity/commit/e8c3430

After that, we removed `extractCritical` because both labels and source maps don't work with that approach. Now it's up to the `@frontity/amp` package add `extractCritical` again for AMP sites. This is where I removed that approach: https://github.com/frontity/frontity/commit/d5af653

Looking at the new docs it seems to me that we don't need to do a separate `hydrate()` on the ids anymore, as they are present in the `data-emotion-css` attribute of the `style` tag:

```js
<style data-emotion-css="${ids.join(' ')}">${css}</style>
```

More info about the `extractCritical` approach on https://emotion.sh/docs/ssr#advanced-approach

### PR DESCRIPTION
**Pull request by cristianbote**
**What**:

Move the critical css to head to pave the way for the amp package.
**Why**:

Amp needs to have the css extracted in the head.
**How**:

Using emotion's server package and the nifty render, template and App apis.
**Tasks**:

- [x] Code
- [x] TSDocs
- [x] TypeScript
- [x] Update other packages
- [x] Unit tests
- [x] TypeScript tests
- [x] Add a changeset (with link to its [Feature Discussion](https://community.frontity.org/c/33) if it exists)

**Unrelated Tasks**

- [ ] End to end tests
- [ ] Update starter themes
- [ ] Open an issue for this feature in the [docs repo](https://github.com/frontity/docs/wiki/Code-Releases)
- [ ] Update community discussions

**Additional Comments**

### CODE REVIEW

File: packages/amp/src/server.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import { CacheProvider } from "frontity";
+import createEmotionServer from "@emotion/server/create-instance";
+import createCache from "@emotion/cache";
+import config, { CACHE_KEY } from "./config";
+
+export default {
```
Comment (michalczaplinski) [inline packages/amp/src/server.tsx]: Could we add the type here?

```js
const ampServer: AMP = { ... };

export default ampServer;
```

File: packages/amp/src/client.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { CacheProvider } from "frontity";
+import createCache from "@emotion/cache";
+import config, { CACHE_KEY } from "./config";
+
+export default {
```
Comment (michalczaplinski) [inline packages/amp/src/client.tsx]: Could we add the type here?

```js
const ampClient: AMP = { ... };

export default ampClient;
```

**Review by michalczaplinski — CHANGES_REQUESTED** [review]
Hey @cristianbote I think the code is great! 👍 

I've only suggested some minor changes, mostly to the tests. Just FYI, I guess it's better if you add the "suggested" changes locally, because if you merge them on github, e.g. the snapshots for the tests won't be generated.

I've also noticed that the types in `@frontity/types/src/frontity` should contain the types for `libraries.frontity.App`, `template` & `render`. Could you add them as well? Here's what I mean: 

https://user-images.githubusercontent.com/5417266/108001674-195c9f00-6fbb-11eb-9a5d-f4f03dcad401.mov

Then, you could change the type of `Packages` in https://github.com/frontity/frontity/blob/amp-package-move-the-css-to-head/packages/amp/types.ts#L55 to include those updated Frontity types like

```ts
export type Packages = MergePackages<AMP, Source, Frontity>;
```

File: packages/amp/src/__tests__/amp.client.test.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      frontity: {
+        App: jest.fn(),
+      },
+    };
+  });
+
+  test("The initial App is overwritten with the Amp one", () => {
+    const previousApp = store.libraries.frontity.App;
+    store.actions.amp.beforeCSR();
+
```
[the remark this answers is not in the document]
Comment (cristianbote) [reply packages/amp/src/__tests__/amp.client.test.ts]: You are correct, I could do better here tests wise. At least making sure the output is gonna stay the same. 👍

File: packages/amp/src/__tests__/amp.server.test.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      },
+    };
+  });
+
+  test("The initial values are overwritten with the Amp ones", () => {
+    const previousRender = store.libraries.frontity.render;
+    const previousTemplate = store.libraries.frontity.template;
+
+    store.actions.amp.beforeSSR();
+
```
[the remark this answers is not in the document]
Comment (cristianbote) [reply packages/amp/src/__tests__/amp.server.test.ts:42]: Indeed we can! Will update the tests. Thank you!

File: packages/amp/src/server.tsx (hunk 1 above)
Comment (cristianbote) [reply packages/amp/src/server.tsx]: oh, oops! Yeah 😅

File: packages/amp/src/client.tsx (hunk 2 above)
Comment (cristianbote) [reply packages/amp/src/client.tsx]: Yuppers!

**Comment by cristianbote** [comment]
Hey @michalczaplinski! 😃  Thank you so much for your review! Indeed, the types for `libraries.frontity` are not good enough. Also gonna update the tests to include their output.

**Comment by cristianbote** [comment]
@michalczaplinski did the changes you suggested. Let me know what you think. Thank you!

**Comment by michalczaplinski** [comment]
This is just a comment that I forgot to mention earlier: 

I think that we could add the e2e tests using the `amphtml-validator` later on when we have the rest of the missing pieces of the AMP package. I have the boilerplate for the e2e-amp package ready in my PR: https://github.com/frontity/frontity/pull/718

File: projects/mars-theme/frontity.settings.js [hunk 5]
```diff
@@ -8,6 +8,7 @@ const settings = {
     },
   },
   packages: [
+    "@frontity/amp",
```
Comment (michalczaplinski) [inline projects/mars-theme/frontity.settings.js]: Hmm, do you want to add the AMP package for some specific reason to mars-theme or was it just something you did for testing?

File: projects/mars-theme/package.json [hunk 6]
```diff
@@ -20,6 +20,7 @@
     "@frontity/mars-theme": "^1.3.1",
     "@frontity/tiny-router": "^1.1.0",
     "@frontity/wp-source": "^1.5.0",
+    "@frontity/amp": "0.0.1",
```
Comment (michalczaplinski) [inline projects/mars-theme/package.json]: Same as above: was it just for testing or did you want to add amp to mars-theme for some specific reason? :)

**Review by michalczaplinski — CHANGES_REQUESTED** [review]
Okay, everything looks great @cristianbote sans those two changes in mars-theme. Just let me know - I've got my finger ready on the "approve" button 😄

File: projects/mars-theme/frontity.settings.js (hunk 5 above)
Comment (cristianbote) [reply projects/mars-theme/frontity.settings.js]: Gaah! Totally forgot to remove them sorry! On it right now!

File: projects/mars-theme/package.json (hunk 6 above)
Comment (cristianbote) [reply projects/mars-theme/package.json]:
```suggestion
```
Does this work?

File: projects/mars-theme/frontity.settings.js (hunk 5 above)
Comment (cristianbote) [reply projects/mars-theme/frontity.settings.js]:
```suggestion
```
Remove it!

**Review by michalczaplinski — APPROVED** [review]
🚀

**Comment by cristianbote** [comment]
So this one has been reviewed twice over 😄  once as the amp template, render to static markup and once as the extract the critical css. I believe it's ready.

**Pull request merged**